### DISCUSSION TITLE
Hardware Spec for Netbox

### DISCUSSION THREAD

**Discussion by Zeawiel** [netbox-community/netbox#11657] [Q&A]
Hello,

I tried to find the hardware spec for the VM on which is installed Netbox.
There is no such thing on the documentation. I try to search on discussion or ticket with no luck.

At this point, we are using Netbox quite heavily (5k IP, 9K cables, 5k devices) with only 2 CPU and 4GB of ram. Which is quite impressive. But it starts to slow down a bit.

Does anyone already made a benchmark or have guideline to know what is the recommended spec at different level of usage?

Best regards !

**Comment by candlerb** [comment]
No, there is no such guide, because it depends entirely on your usage patterns - whether you are hitting the REST API repeatedly, whether you have hundreds of users using the web UI, and so on.

You'll need to monitor the resource usage within your VM, and adjust as appropriate to match your usage pattern.

You may be able to improve response time with your existing VM by increasing the number of gunicorn threads, and/or doing some postgres tuning.

You didn't say what version of Netbox you're using.  If the issue is around the speed of response to global search queries, then migrating to Netbox 3.4 may help (there's a new global search framework).  If the issue is with the speed of displaying results tables, then try to limit your users to 25 or 50 results per page.

**Comment by Zeawiel** [comment]
Hi, thank you for the feedback.

I just uprade to 3.4.3 just before the discussion. 
It seems that my spec is a little underestimated. I'll push to 4 CPU and 8GB of RAM. It should be good.

I don't know anything about gunicorn config, I looked in the doc, and at this moment I have 5 workers (2* 2 cpu +1) which seems ok with the doc.
For the thread I don't know what to do with the value.

I think I'll will wait for the VM spec upgrade an see (and change the gunicorn config to 2*4+1=9 workers).

**Reply by candlerb** [reply]
It depends what you're bottlenecked on.  If you're bottlenecked on external I/O - including waiting for database queries to respond - then more gunicorn threads may be helpful, so that more concurrent CPU work can be done while waiting for responses.

But if the database itself is the bottleneck, either because of CPU or disk I/O, then you'll need to focus there instead.

For looking at real-time resource utilisation, [netdata](https://github.com/netdata/netdata) is very good.  Ignore the netdata cloud stuff; you just need to [install](https://learn.netdata.cloud/docs/agent/packaging/installer/methods/kickstart) netdata inside the Netbox VM, and point your web browser at http://x.x.x.x:19999/ and you'll get real-time metrics with 1 second resolution.

**Comment by LHBL2003** [comment]
Is this set in the /opt/netbox/gunicorn.py or in the /opt/netbox/contrib/gunicorn.py file?

My VM has 8 cores per socket (8 cores / 1 socket).
The VM is hosted on an ESXi with SAN Storrage.

I have set the following values :
```
workers = 17 #instead of 5
threads = 15 #instead of 3
```
I then restarted the computer. I did not notice any difference at first. 

**Question:**
With Microsoft SQL Server you can say how much memory it is allowed to use. Is it possible to do the same here, so that the read speed is increased, e.g. because a lot of data is buffered in the cache? There should be no lack of RAM.

Thanks for any tips.

**Comment by candlerb** [comment]
> Is this set in the /opt/netbox/gunicorn.py or in the /opt/netbox/contrib/gunicorn.py file?

`/opt/netbox/gunicorn.py`, assuming you're using the standard `/etc/systemd/system/netbox.service` which references it:

```
ExecStart=/opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
                                                                                                           ^^^^^^^^^^^^^^^^^^^^^^^
```

> I then restarted the computer. I did not notice any difference at first.

What differences were you looking for? If you use

```
ps -eLF | grep gunicorn
```

then you should see the threads.

As for RAM usage: Linux automatically uses whatever RAM is available for use as a disk cache, which will speed up repeated I/O to the same blocks. For tuning of Postgres, I suggest that's out of scope for this list; you'd be better off looking at dedicated Postgres resources.
